src/System.IO.FileSystem/src/System/IO/Win32FileStream.cs

```diff
@@ -468,10 +469,13 @@ protected override void Dispose(bool disposing)
                 _canRead = false;
                 _canWrite = false;
                 _canSeek = false;
-                // Don't set the buffer to null, to avoid a NullReferenceException
-                // when users have a race condition in their code (ie, they call
-                // Close when calling another method on Stream like Read).
-                //_buffer = null;
+
+                if (_buffer != null)
+                {
+                    ArrayPool<byte>.Shared.Return(_buffer, clearArray: true);
```

Comment:
Is there any established precedent? e.g. SqlClient pools its connections and also has async functions and essentially you are interested in guarding against someone doing something like:
```csharp
List<Task> taskList = new List<Task>();

for (var i = 0; i < 1000; i++){

    using (var db = _dbProviderFactory.CreateConnection())
    using (var cmd = db.CreateCommand())
    {
        ...
        cmd.Parameters.Add(id);

        db.ConnectionString = _connectionString;
        await db.OpenAsync();
        using (var rdr = await cmd.ExecuteReaderAsync())
        {
            taskList.Add(Task.Run(async => {
                while(await rdr.ReadAsync())
                {
                    var Id = rdr.GetInt32(0);
                    var RandomNumber = rdr.GetInt32(1);
                }
            });
        }
    }
}

await Task.WhenAll(taskList);
```